### PR TITLE
Add server and flagger

### PR DESCRIPTION
**Pull request by samfishman**
This PR contains the business logic for the feature-flagging service and defines the API endpoints.

cc @ericavonb @ahsojar

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+	mux := http.NewServeMux()
+	mux.Handle("/public/", http.FileServer(http.Dir("./")))
+	mux.Handle("/admin/flags", admin)
+
+	mux.HandleFunc("/api/flag/create", NewCreateFlagHandler(flagger))
+	mux.HandleFunc("/api/flag/delete/", NewDeleteFlagHandler(flagger))
+	mux.HandleFunc("/api/flag/", NewFlagInfoHandler(flagger))
+	mux.HandleFunc("/api/user/enabled/", NewIsEnabledHandler(flagger))
+	mux.HandleFunc("/api/user/set/", NewSetUserHandler(flagger))
```
Comment (kvigen) [inline server/server.go]: we have some established patterns for writing http services in Go that might be worth looking at for this. I think https://github.com/Clever/transcripts-api might be a good place to look for examples. Let me know if you have any questions.

File: flagger/postgres.go [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+import (
+	"database/sql"
+	"log"
+	"math/rand"
+	"time"
+
+	"github.com/lib/pq"
+)
+
+type Postgres struct {
```
Comment (kvigen) [inline flagger/postgres.go]: what was the rationale for choosing Postgres over Mongo? I'm wary of expanding our database usage without a really compelling reason

Comment (samfishman) [inline flagger/postgres.go]: Mostly because we wanted to be able to spin up a new db quickly at the hackathon, and using RDS seemed like the best way to go (DynamoDB was also considered).

File: server/server.go (hunk 1 above)
Comment (samfishman) [inline server/server.go]: How gorilla/mux compare to httprouter? I have a branch at https://github.com/Clever/flag-bearer/tree/http-router that uses httprouter that works well.

Comment (kvigen) [inline server/server.go]: This is the article we read when we made the decision originally: http://nicolasmerouze.com/guide-routers-golang/. At a high level our thinking was that we valued a bit more functionality over improved performance. I don't have a strong opinion one way or the other, but I do value consistency unless there's a good reason not to use it.

That said, the thing I'm more interested in from the transcripts-api is all the logging / error handling / json marshaling etc.. logic it gets from httpwrapper (https://github.com/Clever/transcripts-api/blob/master/api/api.go#L72). We like the pattern we came up with there, and I would love your feedback on it.

File: flagger/postgres.go (hunk 2 above)
Comment (kvigen) [inline flagger/postgres.go]: Makes sense. Is it worth re-visiting that decision if we're thinking about making this a production service?

Comment (samfishman) [inline flagger/postgres.go]: Definitely worth revisiting, but there are other benefits to Postgres, too – namely, that it can be managed completely by Amazon with RDS.

Comment (nikhilpandit) [inline flagger/postgres.go]: @rgarcia can provide more context on thought process behind considering RDS. I am curious as to learn, why postgres over DynamoDB for this use case?

File: flagger/flagger_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	if err != nil {
+		t.Fatal(fmt.Sprintf("error getting users: %+v", m))
+	}
+	en, ok := users["id0"]
+	if !ok {
+		t.Fatal("feature not noted as disabled for id0")
+	}
+	if en {
+		t.Fatal("feature not disabled for id0")
+	}
```
Comment (rgarcia) [inline flagger/flagger_test.go]: nit: can tighten the scope of some of these variables in this function

```
if err := ...; err != nil {
... 
}
```
and

```
if val, err := ...; err != nil {
...
} else if !val {
...
}
```

**Comment by azylman** [comment]
@kshen0 What's the state of this? If it's not relevant anymore, should we close it?

**Comment by nikhilpandit** [comment]
cc @jacobsimon - was this PR merged into your chemistry hackathon branch before we switched to the other repo? if so, is it ok to merge this?

**Comment by kshen0** [comment]
@azylman this was never reviewed due to time constraints. Given the choice between closing this and interrupting regularly scheduled programming to review and merge this, I choose the latter.

**Comment by azylman** [comment]
@kshen0 Do you want to take over for Sam and respond to PR comments?

**Comment by kshen0** [comment]
@azylman Yeah, I can try to sneak this in

**Comment by samfishman** [comment]
I could respond to this if you want @kshen0 :grinning:. It never got a proper review, though.

**Comment by kshen0** [comment]
@samfishman I'm happy to review if you want to address comments. No pressure though, we can also handle it at Clever HQ

**Comment by samfishman** [comment]
@kshen0 I think I should have time for it!

File: flagger/flagger_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	err := m.CreateFeature("tester", "desc")
+	if err != nil {
+		t.Fatal(fmt.Sprintf("error creating feature: %+v", m))
+	}
+
+	err = m.EnableFeature("id0", "tester")
+	if err != nil {
+		t.Fatal(fmt.Sprintf("error enabling feature: %+v", m))
+	}
+	en, err := m.IsEnabled("id0", "tester")
```
Comment (kshen0) [inline flagger/flagger_test.go]: recommend just using full word `enabled`

File: flagger/README.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+```go
+func (m *Memory) IsEnabled(id string, feature string) (bool, error)
+```
+
+#### func (*Memory) ListFeatures
+
+```go
+func (m *Memory) ListFeatures() ([]FeatureInfo, error)
+```
```
Comment (kshen0) [inline flagger/README.md]: Add DB storage type to readme

File: flagger/memory.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+func (m *Memory) IsEnabled(id string, feature string) (bool, error) {
+	if feature, ok := m.Features[feature]; ok {
+		if is_enabled, ok := feature.UseList[id]; ok {
+			return is_enabled, nil
+		} else if feature.UsePercent {
+			is_enabled = rand.Float32() <= feature.WhitelistPercent
+			feature.UseList[id] = is_enabled
+			return is_enabled, nil
```
Comment (kshen0) [inline flagger/memory.go]: The original intent was to implement this by hashing some key like `id:feature_id` and picking the first n% of resulting hashes. This has a few benefits like being deterministic and allowing users of flag bearer to know ahead of time which users are being enabled on the feature. I'd suggest implementing `feature.UsePercent` in a separate PR in this way. Thoughts?

File: flagger/memory.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+			Description: description,
+			CreatedOn:   time.Now(),
+			UsePercent:  false,
+		},
+		UseList: map[string]bool{},
+	}
+	return nil
+}
+
+func (m *Memory) CreatePercentageFeature(name, description string, pct float32) error {
```
Comment (kshen0) [inline flagger/memory.go]: Since you already did the work to set up this interface, I wouldn't be opposed to simply making these no-ops if we opt to separate Percentage Features to another PR.

File: flagger/postgres.go (hunk 2 above)
Comment (kshen0) [inline flagger/postgres.go]: re-pinging @kvigen - @ericavonb suggests Dynamo might make more sense since we don't need to do much joining, only key-value store (which I agree with)

File: flagger/postgres.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+			return ErrFeatureAlreadyExists
+		} else {
+			log.Println(err)
+		}
+	}
+	return err
+}
+
+func (pg *Postgres) DeleteFeature(name string) error {
+	txn, err := pg.DB.Begin()
```
Comment (kshen0) [inline flagger/postgres.go]: suggest using full variable name `transaction`

File: server/server.go (hunk 1 above)
Comment (kshen0) [inline server/server.go]: I agree with prioritizing consistency unless there's a strong argument for using mux or http-router.

**Comment by kshen0** [comment]
@samfishman I was just reminded that a second hackathon project (from a month or so ago, currently not open sourced) built on this branch of flagbearer. I think the cleanest way forward is to close this PR and apply changes to that repository. Let me know your thoughts.
